src/bin/ion/commands/beta/generate/utils.rs

```diff
@@ -0,0 +1,177 @@
+            Language::Java => "java",
+        }
+    }
+
+    pub fn file_name(&self, name: &str) -> String {
+        match self {
+            Language::Rust => name.to_case(Case::Snake),
+            Language::Java => name.to_case(Case::UpperCamel),
+        }
+    }
+}
+
+impl From<&str> for Language {
+    fn from(value: &str) -> Self {
+        match value {
+            "java" => Language::Java,
+            "rust" => Language::Rust,
+            _ => unreachable!("Unsupported programming language: {}, this tool only supports Java and Rust code generation.", value)
+        }
+    }
+}
+
+impl Display for Language {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Language::Rust => {
+                write!(f, "rust")
+            }
+            Language::Java => {
+                write!(f, "java")
+            }
+        }
+    }
+}
+
+/// Represents all the supported templates for code generation.
+/// These templates will be used by [tera] templating engine to render the generated code with appropriate context value.
+/// _Note: These template variants are based on Rust programming language.
+/// [Template::name] provides an appropriate template file name based on given programming language._
+///
+/// [tera]: <https://docs.rs/tera/latest/tera/>
+pub enum Template {
+    Struct, // Represents a template for a Rust struct or Java class
+}
+
+impl Template {
+    /// Returns a string that represent the template file name based on given programming language.
+    pub fn name(&self, language: &Language) -> &str {
+        match language {
+            Language::Rust => "struct",
+            Language::Java => "class",
+        }
+    }
+}
+
+impl TryFrom<Option<&DataModel>> for Template {
+    type Error = CodeGenError;
+
+    fn try_from(value: Option<&DataModel>) -> Result<Self, Self::Error> {
+        match value {
+            Some(DataModel::Value) | Some(DataModel::Sequence) | Some(DataModel::Struct) => {
+                Ok(Template::Struct)
+            }
+            None => invalid_data_model_error(
+                "Can not get a template without determining data model first.",
+            ),
+        }
+    }
+}
+
+/// Represents an Ion schema type which could either be one of the [built-int types] or a user defined type.
+///
+/// [built-in types]: `<https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#built-in-types>`
+// TODO: Add enum variants for missing built-in ISL types.
+pub enum IonSchemaType {
+    Int,
+    String,
+    Symbol,
+    Float,
+    Bool,
+    Blob,
+    Clob,
+    SchemaDefined(String), // A user defined schema type
+}
+
+impl IonSchemaType {
+    /// Maps the given ISL type name to a target type
+    pub fn target_type(&self, language: &Language) -> &str {
+        use IonSchemaType::*;
+        use Language::*;
+        match (self, language) {
+            (Int, Rust) => "i64",
+            (Int, Java) => "int",
```

Comment:
This should be _at least_ a `long`, and probably even a `BigDecimal`.

Reply:
https://github.com/amazon-ion/ion-cli/issues/78